### PR TITLE
Constants at file-level.

### LINKED ISSUE #9671: Allow constants at file-level

### PR DESCRIPTION
**Pull request by chriseth**
fixes #9671 

*scratches head* - this actually seems to work...

TODO:
 - [x] documentation
 - [x] tests
 - [x] circularity checks
 - [x] inline assembly
 - [x] ast export
 - [x] natspec

### CODE REVIEW

File: test/libsolidity/semanticTests/constants/constants_at_file_level_referencing.sol [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    function g() public returns (bytes memory) {
+        return b;
+    }
+
+    function h() public returns (bytes memory) {
+        return bytes(c);
+    }
+
+    function i() public returns (uint, ActionChoices, bytes32) {
+        return (x, choices, st);
```
Comment (leonardoalt) [inline test/libsolidity/semanticTests/constants/constants_at_file_level_referencing.sol]: where are these declared?

Comment (leonardoalt) [reply test/libsolidity/semanticTests/constants/constants_at_file_level_referencing.sol]: ah the other test

File: libsolidity/parsing/Parser.cpp [hunk 2]
```diff
@@ -111,7 +111,21 @@ ASTPointer<SourceUnit> Parser::parse(shared_ptr<Scanner> const& _scanner)
 				nodes.push_back(parseFunctionDefinition(true));
 				break;
 			default:
-				fatalParserError(7858_error, "Expected pragma, import directive or contract/interface/library/struct/enum/function definition.");
+				if (
+					m_scanner->currentToken() == Token::Identifier ||
+					m_scanner->currentToken() == Token::Mapping ||
```
Comment (leonardoalt) [inline libsolidity/parsing/Parser.cpp]: Constant mapping?

File: libsolidity/parsing/Parser.cpp [hunk 3]
```diff
@@ -111,7 +111,21 @@ ASTPointer<SourceUnit> Parser::parse(shared_ptr<Scanner> const& _scanner)
...
+					TokenTraits::isElementaryTypeName(m_scanner->currentToken()) ||
+					(m_scanner->currentToken() == Token::Function && m_scanner->peekNextToken() == Token::LParen)
```
Comment (leonardoalt) [inline libsolidity/parsing/Parser.cpp]: Why function also?

File: libsolidity/parsing/Parser.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: This is just the generic code for variable declaration, I'll probably pull it out.

File: libsolidity/parsing/Parser.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: Same reason.

File: test/libsolidity/semanticTests/constants/constants_at_file_level_referencing.sol (hunk 1 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/constants/constants_at_file_level_referencing.sol]: This file already creates a segfault at an earlier time in the processing, but the other test works fine.

**Comment by leonardoalt** [comment]
Does the grammar count as documentation?

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `a2d2479085b1aa094e2ced79e50497ff765f6e39`:
```
Error: Trailing whitespace found:
 test/cmdlineTests/recovery_ast_empty_contract/err:5: | 
 test/cmdlineTests/recovery_ast_empty_contract/err:11: | 

```
Please check that your changes are working as intended.

File: libsolidity/parsing/Parser.cpp [hunk 4]
```diff
@@ -111,7 +111,17 @@ ASTPointer<SourceUnit> Parser::parse(shared_ptr<Scanner> const& _scanner)
 				nodes.push_back(parseFunctionDefinition(true));
 				break;
 			default:
-				fatalParserError(7858_error, "Expected pragma, import directive or contract/interface/library/struct/enum/function definition.");
+				// Constant variable.
+				if (variableDeclarationStart())
+				{
+					VarDeclParserOptions options;
+					options.isStateVariable = true;
```
Comment (leonardoalt) [inline libsolidity/parsing/Parser.cpp]: Why are file level constants considered state variables?

File: liblangutil/SourceReferenceFormatterHuman.cpp [hunk 5]
```diff
@@ -105,7 +105,7 @@ void SourceReferenceFormatterHuman::printSourceLocation(SourceReference const& _
 
 	if (!_ref.multiline)
 	{
-		auto const locationLength = static_cast<size_t>(_ref.endColumn - _ref.startColumn);
+		size_t const locationLength = static_cast<size_t>(_ref.endColumn - _ref.startColumn);
```
Comment (leonardoalt) [inline liblangutil/SourceReferenceFormatterHuman.cpp:108]: Why? It's already clear what the type is

File: test/libsolidity/semanticTests/constants/constant_string_at_file_level.sol [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+    function g() public returns (bytes memory) {
+        return b;
+    }
+
+    function h() public returns (bytes memory) {
+        return bytes(c);
+    }
+
+    function i() public returns (uint, ActionChoices, bytes32) {
```
Comment (leonardoalt) [inline test/libsolidity/semanticTests/constants/constant_string_at_file_level.sol:22]: This function is unused.

File: test/libsolidity/syntaxTests/constants/constant_natspec_user.sol [hunk 7]
```diff
@@ -0,0 +1,4 @@
+/// Documentation
+uint constant x = 8;
+// ----
+// DocstringParsingError 6546: (0-18): Documentation tag @notice not valid for non-public state variables.
```
Comment (leonardoalt) [inline test/libsolidity/syntaxTests/constants/constant_natspec_user.sol]: Why is it complaining about notice?

File: libsolidity/parsing/Parser.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: This is just for the parser. This allows to parse the 'constant' keyword. I could also heavily change that method, but I thought it better to add extensive checks in the analysis phase later.

Should I rename the option field to `isStateOrFileLevelVariable`?

File: test/libsolidity/syntaxTests/unexpected.sol [hunk 8]
```diff
@@ -1,3 +1,3 @@
 unexpected
 // ----
-// ParserError 7858: (0-10): Expected pragma, import directive or contract/interface/library/struct/enum/function definition.
+// ParserError 2314: (11-11): Expected identifier but got end of source
```
Comment (leonardoalt) [inline test/libsolidity/syntaxTests/unexpected.sol]: The error was better before

File: liblangutil/SourceReferenceFormatterHuman.cpp (hunk 5 above)
Comment (chriseth) [reply liblangutil/SourceReferenceFormatterHuman.cpp:108]: It is another indirection away.

**Review by leonardoalt — CHANGES_REQUESTED** [review]
Can you also add tests with file level constants with the same name but from different sources?

File: test/libsolidity/syntaxTests/constants/constant_natspec_user.sol (hunk 7 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/constants/constant_natspec_user.sol]: Because the default is notice.

File: libsolidity/parsing/Parser.cpp (hunk 4 above)
Comment (leonardoalt) [reply libsolidity/parsing/Parser.cpp]: It's fine, just wanted to check.

File: test/libsolidity/syntaxTests/unexpected.sol (hunk 8 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/unexpected.sol]: I know, I tried to improve it but it's rather difficult. If we allow `a.b.c.EnumName constant variableName = a.b.c.EnumName.EnumMember`, then this can always happen. I can add a "hack" that at least detects the case of `<identiifer> <EOS>`.

Comment (leonardoalt) [reply test/libsolidity/syntaxTests/unexpected.sol]: If it's a hack or too much work then I think it's fine like this. These cases are [hopefully] rare anyway

Comment (chriseth) [reply test/libsolidity/syntaxTests/unexpected.sol]: Updated it, it's not too bad.

Comment (leonardoalt) [reply test/libsolidity/syntaxTests/unexpected.sol]: Ok!

**Review by Marenz — DISMISSED** [review]
As far as I understood things, they look good.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 9]
```diff
@@ -1753,18 +1753,24 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 	{
 		Type::Category category = _memberAccess.annotation().type->category();
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:1754]: Added this just now.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 10]
```diff
@@ -1871,6 +1871,35 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 			solAssert(false, "");
 		break;
 	}
+	case Type::Category::Module:
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1874]: Added this just now.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 11]
```diff
@@ -2130,6 +2159,10 @@ void IRGeneratorForStatements::endVisit(Identifier const& _identifier)
 	{
 		// no-op
 	}
+	else if (dynamic_cast<ImportDirective const*>(declaration))
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:2162]: Added this just now.

File: libsolidity/parsing/Parser.cpp [hunk 12]
```diff
@@ -111,7 +111,17 @@ ASTPointer<SourceUnit> Parser::parse(shared_ptr<Scanner> const& _scanner)
 				nodes.push_back(parseFunctionDefinition(true));
 				break;
 			default:
-				fatalParserError(7858_error, "Expected pragma, import directive or contract/interface/library/struct/enum/function definition.");
+				// Constant variable.
+				if (variableDeclarationStart() && m_scanner->peekNextToken() != Token::EOS)
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:115]: This is the parser fix.

**Comment by chriseth** [comment]
Found another problem, but hopefully the last.

File: libsolidity/analysis/PostTypeChecker.cpp [hunk 13]
```diff
@@ -162,6 +170,15 @@ struct ConstStateVarCircularReferenceChecker: public PostTypeChecker::Checker
 		return true;
 	}
 
+	bool visit(MemberAccess const& _memberAccess) override
```
Comment (chriseth) [inline libsolidity/analysis/PostTypeChecker.cpp:173]: Added this.

File: test/libsolidity/syntaxTests/constants/cross_file_cyclic_modules.sol [hunk 14]
```diff
@@ -0,0 +1,15 @@
+==== Source: a ====
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/constants/cross_file_cyclic_modules.sol:1]: Newly added.

File: test/libsolidity/semanticTests/constants/same_constants_different_files.sol [hunk 15]
```diff
@@ -0,0 +1,26 @@
+==== Source: s1.sol ====
```
Comment (chriseth) [inline test/libsolidity/semanticTests/constants/same_constants_different_files.sol:1]: Test requested by @leonardoalt

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 16]
```diff
@@ -1753,18 +1753,24 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
...
 		solAssert(
+			dynamic_cast<VariableDeclaration const*>(_memberAccess.annotation().referencedDeclaration) ||
+			dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration) ||
 			category == Type::Category::TypeType ||
-			category == Type::Category::Module ||
-			category == Type::Category::Function,
```
Comment (leonardoalt) [inline libsolidity/codegen/ExpressionCompiler.cpp:1758]: Was this buggy? (Since it was replaced by a similar check on `referencedDeclaration` instead of `_memberAccess`)

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 17]
```diff
@@ -1871,6 +1871,35 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
...
+			category == Type::Category::TypeType ||
+			category == Type::Category::Module,
+			""
+		);
+		if (auto variable = dynamic_cast<VariableDeclaration const*>(_memberAccess.annotation().referencedDeclaration))
+		{
+			solAssert(variable->isConstant(), "");
+			handleVariableReference(*variable, static_cast<Expression const&>(_memberAccess));
+		}
+		else if (auto const* function = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration))
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1889]: This seems unrelated. Is it adding support to accessing free functions via the module?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 18]
```diff
@@ -1871,6 +1871,35 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
...
+		{
+			auto funType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type);
+			solAssert(function && function->isFree(), "");
+			solAssert(function->functionType(true), "");
+			solAssert(function->functionType(true)->kind() == FunctionType::Kind::Internal, "");
+			solAssert(funType->kind() == FunctionType::Kind::Internal, "");
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1895]: How's this check different from the one in the line above?

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 16 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1758]: I just feel safer checking the kind of what is referenced instead of the type of its value. Since the type of a variable's value can be just about anything, I changed it.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 17 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1889]: Yes exactly, it produced an ICE before.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 18 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1895]: Better safe than sorry.

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1895]: One check is on the type of the expression, the other on the type created from the function definition.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 17 above)
Comment (leonardoalt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1889]: Do we have tests for that?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 18 above)
Comment (leonardoalt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1895]: Ok

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 16 above)
Comment (leonardoalt) [reply libsolidity/codegen/ExpressionCompiler.cpp:1758]: Ok

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 17 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1889]: Added one: https://github.com/ethereum/solidity/pull/9892/files#diff-6797bc7588edb0c5dd1a3ccad7d9f07f

File: libsolidity/parsing/Parser.cpp (hunk 4 above)
Comment (ekpyron) [reply libsolidity/parsing/Parser.cpp]: Maybe renaming it wouldn't be the worst of ideas? I guess it could potentially spare us quite some headaches in the future...

File: test/libsolidity/syntaxTests/constants/constant_natspec_user.sol (hunk 7 above)
Comment (ekpyron) [inline test/libsolidity/syntaxTests/constants/constant_natspec_user.sol]: The message is not too nice now, but maybe not worth the effort of changing...

File: libsolidity/parsing/Parser.cpp (hunk 4 above)
Comment (ekpyron) [reply libsolidity/parsing/Parser.cpp]: Also not sure if a ``isFileLevelVariable`` field in ``VarDeclParserOptions`` instead would have been that bad... then ``parseVariableDeclaration`` could just reject docstrings, have a better error message for the function type thing, and reject visibility and override...
But yeah - maybe doing all of that in analysis has advantages - one could argue either way I guess and both is fine with me (and maybe not worth changing all that now anyways).

But at least renaming the field would be nice.

File: test/libsolidity/syntaxTests/constants/constant_override.sol [hunk 19]
```diff
@@ -0,0 +1,3 @@
+uint constant override x;
+// ----
+// TypeError 4266: (0-24): Uninitialized "constant" variable.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/constants/constant_override.sol]: This test should maybe initialize it? Does that error out? This is not the error I would expect in this test in any case.

File: test/libsolidity/syntaxTests/constants/constant_with_visibility.sol [hunk 20]
```diff
@@ -0,0 +1,3 @@
+uint public x = 7;
+// ----
+// DeclarationError 8342: (0-17): Only constant variables are allowed at file level.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/constants/constant_with_visibility.sol]: Well, ok, but what about ``uint constant public x = 7;``?

File: test/libsolidity/syntaxTests/constants/redefinition_public_constant.sol [hunk 21]
```diff
@@ -0,0 +1,4 @@
+uint constant public c = 7;
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/constants/redefinition_public_constant.sol]: Shouldn't this already be an error due to the ``public``? Or what does ``public`` or non-``public`` mean for constants?

File: test/libsolidity/semanticTests/constants/same_constants_different_files.sol [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+==== Source: s2.sol ====
+
+import {a as b, fre} from "s1.sol";
+import "s1.sol" as M;
+
+uint256 constant a = 13;
+
+contract C {
+    function f() public returns (uint, uint, uint) {
+        return (a, fre(), M.a);
```
Comment (ekpyron) [inline test/libsolidity/semanticTests/constants/same_constants_different_files.sol]: Why not also test ``b`` while we're at it?

**Review by ekpyron — CHANGES_REQUESTED** [review]
Maybe I'm being a bit paranoid here... but I'd like to see some more tests yet :-). I commented on some existing tests, but also ``uint[] constant memory x = ...`` etc. might be good (it will be a parser error, I see that, but maybe better to be excessive about tests than to lack them in the end...).

Also maybe a test of a file-level function and a file-level constant having the same name or did I overlook that?

File: test/libsolidity/syntaxTests/constants/constant_override.sol (hunk 19 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/constants/constant_override.sol]: Good catch!

File: test/libsolidity/syntaxTests/nameAndTypeResolution/free_and_constant.sol [hunk 23]
```diff
@@ -0,0 +1,4 @@
+uint constant c = 7;
+function c() returns (uint) {}
+// ----
+// DeclarationError 2333: (21-51): Identifier already declared.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/free_and_constant.sol:4]: I hope this means that *all* conflicts of top-level construct are covered now. Probably, the function case was just special due to overload resolution and stuff, so it should be fine... theoretically we could have 42 distinct test cases like this to really cover everything explicitly, but I guess it's fine (and granted a handful of those cases probably already have other tests) :-).

**Review by ekpyron — APPROVED** [review]
With these "allow stuff at top-level" changes it's always hard to tell, if one is missing anything, because the main part of the work is done by the declaration registration, resp. name and type resolution which hardly needs changes for it, but wasn't written with all that in mind...
But I go out on a limb and say this is probably fine now :-).
